src/libostree/ostree-bootloader-zipl.c

```diff
@@ -78,8 +85,197 @@ _ostree_bootloader_zipl_write_config (OstreeBootloader  *bootloader,
+                                         gchar **options,
+                                         GCancellable *cancellable,
+                                         GError **error)
+{
+  g_autoptr (GPtrArray) configs = NULL;
+  if ( !_ostree_sysroot_read_boot_loader_configs (self->sysroot, bootversion, &configs, cancellable, error))
+    return glnx_prefix_error (error, "s390x SE: loading bls configs");
+
+  if (!configs || configs->len == 0)
+    return glnx_throw (error, "s390x SE: no bls config");
+
+  OstreeBootconfigParser *parser = (OstreeBootconfigParser *) g_ptr_array_index (configs, 0);
+  const gchar *val = NULL;
+
+  val = ostree_bootconfig_parser_get (parser, "linux");
+  if (!val)
+    return glnx_throw (error, "s390x SE: no \"linux\" key in bootloader config");
+  *vmlinuz = g_build_filename ("/boot", val, NULL);
+
+  val = ostree_bootconfig_parser_get (parser, "initrd");
+  if (!val)
+    return glnx_throw (error, "s390x SE: no \"initrd\" key in bootloader config");
+  *initram = g_build_filename ("/boot", val, NULL);
+
+  val = ostree_bootconfig_parser_get (parser, "options");
+  if (!val)
+    return glnx_throw (error, "s390x SE: no \"options\" key in bootloader config");
+  *options = g_strdup(val);
+
+  return TRUE;
+}
+
+static gboolean
+_ostree_secure_execution_luks_key_exists (void)
+{
+  return (access(SECURE_EXECUTION_LUKS_ROOT_KEY, F_OK) == 0 &&
+          access(SECURE_EXECUTION_LUKS_CONFIG, F_OK) == 0);
+}
+
+static gboolean
+_ostree_secure_execution_enable_luks(gchar* initram, GError **error)
+{
+  static const char *cmd = "/usr/libexec/libostree/s390x-se-luks-gencpio";
+  const char *const argv[] = {cmd, initram, SECURE_EXECUTION_INITRD_IMAGE, NULL};
+
+  g_autofree gchar *out = NULL;
+  g_autofree gchar *err = NULL;
+  int status = 0;
+  if (!g_spawn_sync (NULL, (char**)argv, NULL, G_SPAWN_SEARCH_PATH,
+                     NULL, NULL, &out, &err, &status, error))
+    return glnx_prefix_error(error, "s390x SE: spawning %s", cmd);
+
+  if (!g_spawn_check_exit_status (status, error))
+    {
+      g_printerr("s390x SE: `%s` stdout: %s\n", cmd, out);
+      g_printerr("s390x SE: `%s` stderr: %s\n", cmd, err);
+      return glnx_prefix_error(error, "s390x SE: `%s` failed", cmd);
+    }
+
+  sd_journal_print(LOG_INFO, "s390x SE: luks key added to initrd");
+  return TRUE;
+}
+
+static gboolean
+_ostree_secure_execution_generate_sdboot (gchar *vmlinuz,
+                                          gchar *initram,
+                                          gchar *options,
+                                          GPtrArray *keys,
+                                          GError **error)
+{
+  if (!vmlinuz || !initram || !options || !keys)
+    return glnx_throw (error, "s390x SE: vmlinuz, initram, options, keys are missing");
+
+  g_autoptr(GString) cmd   = NULL;
+  g_autofree char *cmdline = g_strdup ("/tmp/sd_boot.parmfile.XXXXXX");
+  glnx_autofd int fd       = g_mkstemp (cmdline);
```

Comment:
Hmm, I don't think we use `g_mkstemp` much. There's a suite of better tmpfile APIs in libglnx. Though ahh OK, you need to pass it as a path to `genprotimg`.

I think it'd work to use `glnx_open_anonymous_tmpfile` for this and then pass the path as `/proc/$pid/fd/$fd`.